client/src/optimizing_client.rs

```diff
@@ -0,0 +1,250 @@
+//! The `optimizing_client` module is a super-client which
+//! keeps stats about a list of clients and uses the best one.
+
+use crate::thin_client::ThinClient;
+use core::cmp::min;
+use rand::{thread_rng, Rng};
+use solana_sdk::client::{AsyncClient, Client, SyncClient};
+use solana_sdk::hash::Hash;
+use solana_sdk::instruction::Instruction;
+use solana_sdk::message::Message;
+use solana_sdk::pubkey::Pubkey;
+use solana_sdk::signature::{Keypair, Signature};
+use solana_sdk::timing::duration_as_ms;
+use solana_sdk::transaction::{self, Transaction};
+use solana_sdk::transport::Result as TransportResult;
+use std::collections::VecDeque;
+use std::io;
+use std::sync::RwLock;
+use std::time::Duration;
+use std::time::Instant;
+
+/// An object for querying and sending transactions to the network.
+pub struct OptimizingClient {
```

Comment:
How about calling this something like `ValidatorClients` with one method `closest(&self) -> &ThinClient`?

Reply:
Well, I think the behavior is randomly testing a set of nodes so closest to me sounds like a misnomer. Isn't it also nice to be able to switch out this with the other clients like BankClient and the regular client?

Reply:
It's a nice abstraction, but looks like a leaky one. It doesn't seem like one would want to dynamically switch clients at each and every client call. You'd want to pick a closest client, use it for a bit, and then perhaps switch.  Also, is it worth maintaining a wrapper that needs to be updated any time that trait changes?

Reply:
Also, another way to do this is within ThinClient. Instead of swapping clients you can swap socket addresses.

Reply:
`client()` instead of `closest()`?

Reply:
Sure, switching each call might be excessive, but this gives you the flexibility to try any strategy.  Maintaining the interface is a bit annoying, I guess in your suggestion you would have some kind of experimentation period maybe on creation and then periodically, then the user would call this client() method to get the best one.

Reply:
The client may have better information about when to experiment, then in the shim design you need to give some hints how to operate.

Reply:
Yes, exactly. My suggestion is to give the user the control until we start seeing a pattern of every user using it the same way. At that point (which may be never) we formalize it with an abstraction.

Reply:
I used switching sockets in thin_client.